### PR TITLE
feat(realtime): add input guardrails for RealtimeAgent and RealtimeRunConfig

### PR DESCRIPTION
**Pull request by Skyline-9**
### Summary

Adds **input guardrails** to the realtime API, bringing it closer to parity with the non-realtime `Agent`/`Runner`, which already supports `input_guardrails`. Realtime today only supports output guardrails (`RealtimeAgent.output_guardrails` / `RealtimeRunConfig["output_guardrails"]`); there is no first-class way to screen the user's transcribed input.

What changed:

- `RealtimeAgent.input_guardrails` (appended at the end of the dataclass, `default_factory=list`) and `RealtimeRunConfig["input_guardrails"]` (`NotRequired` TypedDict key).
- New `RealtimeInputGuardrailTripped` session event (appended at the end of the `RealtimeSessionEvent` union), mirroring `RealtimeGuardrailTripped` field-for-field but typed to `InputGuardrailResult`.
- `RealtimeSession` runs the combined agent + run-config input guardrails on the **completed user transcript** (`input_audio_transcription_completed`), de-duped by `id()`. It reuses the existing output-guardrail machinery (shared `_guardrail_tasks` set, `_on_guardrail_task_done`, `_cleanup_guardrail_tasks`), so `close()` cancels in-flight tasks. On a trip it emits `input_guardrail_tripped`, forces `response.cancel`, and sends a follow-up user message naming the guardrail.
- Exported from `agents.realtime.__init__` (`__all__`) with an import regression test.
- Docs: `docs/ref/realtime/events.md` renders the new event; `docs/realtime/guide.md` documents the feature and disambiguates it from the existing tool-level "input guardrails on function-tool calls".

The design deliberately mirrors `_run_output_guardrails` (argument order verified against `InputGuardrail.run(self, agent, input, context)`) so the behavior and lifecycle are consistent with what maintainers already review.

### Known limitation (documented, not hidden)

The forced cancel reliably interrupts a response that is already in flight. If a guardrail resolves in the narrow window **before** any response has been created for the tripped turn, the cancel is a no-op and that response may proceed. Eliminating this window cleanly requires response<->user-item correlation at the model layer (for example a `response_id` on turn-started / response-created) so the session can cancel only the tripped turn's response without also cancelling the intentional guardrail-notification response. This limitation is documented in the `RealtimeInputGuardrailTripped` docstring, `RealtimeAgent.input_guardrails`, and the guide rather than papered over with a heuristic that would cancel the wrong response. Scope is also documented: input guardrails run on transcribed audio only; text sent via `send_message` is not screened. Happy to pursue the model-layer correlation as a follow-up if maintainers prefer.

### Test plan

- Added `tests/realtime/test_session.py::TestInputGuardrailFunctionality`, including edge cases:
  - a raising guardrail is skipped and does not crash the shared guardrail task,
  - raising + tripping guardrails together still produce exactly one interrupt with the tripping guardrail named,
  - a second transcription for an already-tripped item is de-duplicated,
  - no guardrail task is created when none are configured.
- Ran the standard verification stack from the repo root:
  - `make format`, `make lint`, `make typecheck` — pass
  - `make tests` (full) — pass (4797 passed, 2 skipped; serial 27 passed, 5 skipped)
  - `make build-docs` — pass (new `RealtimeInputGuardrailTripped` reference resolves clean)

### Issue number

Realtime parity with the non-realtime input-guardrail support. Happy to link the relevant tracking issue.

### Checks

- [x] I've added new tests, if relevant
- [x] I've run `.agents/skills/code-change-verification/scripts/run.sh`
- [x] I've confirmed all verification steps pass (ran `make format`, `make lint`, `make typecheck`, `make tests`, and `make build-docs`)
- [x] If using Codex, I've run `/review` before submitting this PR

### Compatibility notes

Additive. New fields are appended at the end of `RealtimeAgent` (preserving positional compatibility) and are a `NotRequired` config key; the new event is appended at the end of the `RealtimeSessionEvent` union. Sessions with no input guardrails configured create no extra tasks per utterance.

### CODE REVIEW

**Comment by Skyline-9** [comment]
@codex review

**Comment by Skyline-9** [comment]
@codex review

File: src/agents/realtime/session.py [hunk 1]
```diff
@@ -1263,6 +1271,94 @@ async def _run_output_guardrails(self, text: str, response_id: str) -> bool:
...
+            await self._put_event(
+                RealtimeInputGuardrailTripped(
+                    guardrail_results=triggered_results,
+                    message=text,
+                    info=self._event_info,
+                )
+            )
+
+            # Interrupt the model, forcing a cancel of any in-progress response.
+            await self._model.send_event(RealtimeModelSendInterrupt(force_response_cancel=True))
```
[the remark this answers is not in the document]
Comment (Skyline-9) [reply src/agents/realtime/session.py]: Fair point but this is how output guardrail happens too. For the sake of parity, I will acknowledge this as a drawback but not change this, if this is desired, it should be raised in a separate PR

File: src/agents/realtime/session.py [hunk 2]
```diff
@@ -1263,6 +1271,92 @@ async def _run_output_guardrails(self, text: str, response_id: str) -> bool:
...
+                        # Interrupt the model, forcing a cancel of any in-progress response.
+                        await self._model.send_event(
+                            RealtimeModelSendInterrupt(force_response_cancel=True)
+                        )
+
+                        # Send guardrail triggered message.
+                        guardrail_names = [r.guardrail.get_name() for r in triggered_results]
+                        user_input = f"input guardrail triggered: {', '.join(guardrail_names)}"
+                        await self._model.send_event(
+                            RealtimeModelSendUserInput(user_input=user_input)
```
[the remark this answers is not in the document]
Comment (Skyline-9) [reply src/agents/realtime/session.py:1489]: Updated by deleting the offending message in conversation history

**Comment by Skyline-9** [comment]
Hi team,

I have updated the PR to address the recent Codex review comments, rebased it on top of the latest `main`, and verified that all tests/checks pass successfully.

### Key Updates:
1. **Preventing Guardrail Bypass**: When an input guardrail trips, the session now sends a `conversation.item.delete` client event to remove the offending transcribed user input item from the conversation history. This ensures that the replacement response generated in response to the guardrail notification does not see or act on the disallowed input.
2. **Rejection of Sequential/Blocking Guardrails**: Because the Realtime API automatically starts streaming a response concurrently upon user silence (turn-detection), sequential blocking guardrails (`run_in_parallel=False`) cannot prevent response generation and are incompatible. The session now raises a `ValueError` if any input guardrail is configured with `run_in_parallel=False` (fully documented in `docs/realtime/guide.md`).
3. **Rebase & Verification**: Rebased cleanly on top of `upstream/main`. Run `make format`, `make lint`, `make typecheck`, and all tests pass cleanly.

@codex review

**Comment by Skyline-9** [comment]
@codex review
